core/cppgc.rs

```diff
@@ -18,6 +18,80 @@ struct CppGcObject<T: GarbageCollected> {
   member: T,
 }
 
+/// PrototypeChain defines the position of the object in the prototype chain.
+///
+///   A    (index 0)
+///  / \
+/// B   C  (index 1)
+/// |
+/// D      (index 2)
+///
+/// where each B, C and D hold their own prototype chain.
+pub trait PrototypeChain {
+  fn prototype_index() -> Option<usize> {
+    None
+  }
+}
+
+/// A good enough number for the maximum prototype chain depth based on the
+/// usage.
+const MAX_PROTO_CHAIN: usize = 3;
+
+struct DummyT;
+impl GarbageCollected for DummyT {}
+
+/// ErasedPtr is a wrapper around a `v8::cppgc::Ptr` that allows downcasting
+/// to a specific type. Safety is guaranteed by the `tag` field in the
+/// `CppGcObject` struct.
+struct ErasedPtr {
```

Comment:
Actually this abstraction is much nicer than what we had previously. With this it is impossible to accidentially use `T` without checking `ptr.tag` anymore. Great!